krabcake/rs_hello/src/data.rs

```diff
@@ -1,38 +1,93 @@
-use alloc::vec::Vec;
+use alloc::{string::String, vec::Vec};
 use core::fmt;
 
-use crate::{vg_addr, COUNTER, CTX, STACKS};
+use crate::{vg_addr, vg_long, vg_ulong, COUNTER, CTX, STACKS};
 
-#[derive(Copy, Clone, PartialEq, Eq)]
-pub struct Tag(pub u64);
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Tag {
+    Bottom,
+    Id(u64),
+}
 
 impl Tag {
-    pub fn next(self) -> Tag {
-        Tag(self.0 + 1)
+    pub fn next(self) -> Option<Tag> {
+        self.id().map(|id| Tag::Id(id + 1))
+    }
+
+    pub fn id(self) -> Option<u64> {
+        match self {
+            Tag::Id(id) => Some(id),
+            Tag::Bottom => None,
+        }
+    }
+
+    pub fn s(&self) -> String {
+        alloc::format!("{self:?}\0")
+    }
+
+    pub fn to_vg_long(self) -> vg_long {
+        match self {
+            Tag::Id(id) => {
+                // ensure top bits are 0 (and thus can be shifted away safely)
+                assert_eq!(0, id >> 60);
+                (id << 4) as vg_long
+            }
+            Tag::Bottom => 1,
+        }
+    }
+
+    pub fn from_vg_ulong(id: vg_ulong) -> Option<Self> {
+        match id {
+            0 => None,
+            1 => Some(Tag::Bottom),
+            x => {
+                if (x & !0b1111) != 0 {
+                    Some(Tag::Id(x >> 4))
+                } else {
+                    // If the top bits beyond the first four are zero, but the
+                    // whole value is not handled in the earlier match arms,
+                    // then this must be malformatted.
+                    panic!("incorrectly encoded shadow state")
+                }
+            }
+        }
     }
 }
 
-#[derive(Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
 pub enum Item {
     Unique(Tag),
+    SharedReadWrite,
 }
 
 impl fmt::Display for Item {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Item::Unique(tag) => ::core::write!(f, "Unique({})", tag.0),
+            Item::Unique(Tag::Id(id)) => ::core::write!(f, "Unique({})", id),
+            Item::Unique(Tag::Bottom) => ::core::write!(f, "Bottom"),
+            Item::SharedReadWrite => ::core::write!(f, "SharedRW"),
         }
     }
 }
 
 impl Item {
     pub fn num(&self) -> u64 {
         match *self {
-            Item::Unique(Tag(val)) => val,
+            Item::Unique(Tag::Id(val)) => val,
+            Item::Unique(Tag::Bottom) => unimplemented!(),
+            Item::SharedReadWrite => unimplemented!(),
+        }
+    }
+
+    pub fn from_tag(tag: Tag) -> Self {
```

Comment:
I *think* this method is meant to be documented with something like "given PTR = Pointer(LOC, TAG), Item::from_tag(TAG) will return an item IT such that dereferences of PTR will need to pop the stack associated with LOC until it finds an item equivalent to IT."

Reply:
I removed `Item` altogether